include/deal.II/multigrid/mg_matrix.h

```diff
@@ -214,6 +227,22 @@ namespace mg
 
 
 
+  template <typename VectorType>
+  template <typename MatrixType>
+  inline void
+  Matrix<VectorType>::initialize(
+    const MGLevelObject<std::unique_ptr<MatrixType>> &p)
+  {
+    matrices.resize(p.min_level(), p.max_level());
+    for (unsigned int level = p.min_level(); level <= p.max_level(); ++level)
+      {
+        matrices[level] =
+          linear_operator<VectorType>(LinearOperator<VectorType>(), *p[level]);
+      }
```

Comment:
Maybe I misunderstand, but since all the methods are templated on `MatrixType`, can't we pick this up with the existing interface with `MatrixType=std::unique_ptr<SomeActualMatrixType>` and then have an internal helper function that translates the access from a pointer type to a dereference, i.e.,
```c++
       matrices[level] =
          linear_operator<VectorType>(LinearOperator<VectorType>(), maybe_dereference_pointer(p[level]));
```

Reply:
I will try this out.

Reply:
OK. It does work. I would suggest to introduce in the namepace `Utilities` (in base/utilties.h) the following function(s):
```  
  template <typename T>
  T &
  maybe_dereference_pointer(T &p)
  {
      return p;
  }
  
  template <typename T>
  T &
  maybe_dereference_pointer(const std::shared_ptr<T> &p)
  {
      return *p;
  }
  
  template <typename T>
  T &
  maybe_dereference_pointer(const std::unique_ptr<T> &p)
  {
      return *p;
  }
```

Any thoughts on these?

Reply:
I would like this solution, but I am biased. @masterleinad ?

Reply:
I don't like having `maybe` in a function name but something like `get_underlying_value` or similar. I general I prefer this approach over introducing new overloads but I would really like to understand if there is a use case for passing by `const std::unique_ptr<>&` in the first place.